src/types/actions.ts

```diff
@@ -0,0 +1,112 @@
+import {
+  Abi,
+  Account,
+  Address,
+  Chain,
+  SimulateContractParameters,
+  WriteContractParameters,
+} from 'viem'
+import { OpStackL1Contracts } from './opStackContracts'
+
+export type ExtractValidChainIdFromContract<
+  TChain extends Chain | undefined,
+  contractName extends string,
+> = TChain extends Chain
+  ? TChain['contracts'] extends { [key: string]: any }
+    ? TChain['contracts'][contractName] extends { [chainId: number]: Address }
+      ? keyof TChain['contracts'][contractName]
+      : undefined
+    : undefined
+  : undefined
+
+export type ResolveChain<
+  TChain extends Chain | undefined,
+  TChainOverride extends Chain | undefined = undefined,
+> = TChainOverride extends Chain ? TChainOverride : TChain
+
+export type GetL2ChainId<
+  TChain extends Chain | undefined,
+  contractName extends string,
+> = ExtractValidChainIdFromContract<TChain, contractName> extends undefined
+  ? {
+      l2ChainId?: `Contract ${contractName} is not provided on chain. Please add a ${contractName}Address as an arugment`
+    }
+  : // NOTE(Wilson): users will see this as a required arg in the case they are passing optimismPortalAddress
+    // explicitly and the chain has entries at contracts[contractName], e.g. I am using a chain with some
+    // known optimismPortal address but I am sending to a different one. toChainId does not actually get
+    // get used in the code on this path, but I am leaving this as making it optional means NOT passing optimismPortalAddress
+    // and also not passing toChainId is allowed.
+    { l2ChainId: ExtractValidChainIdFromContract<TChain, contractName> }
+
+// actually isn't quite what we want? do we want to let them override
+// contract name might be specified but this specific chain might not be
+// we want to say, if they pass in a toChainId then we need to make sure
+
+export type GetContractAddress<
+  TChain extends Chain | undefined,
+  contractName extends string,
+> = TChain extends Chain
+  ? TChain['contracts'] extends { [key: string]: any }
+    ? TChain['contracts'][contractName] extends { [chainId: number]: Address }
+      ? {
+          [k in `${contractName}Address`]?: Address
+        }
+      : {
+          [k in `${contractName}Address`]: Address
+        }
+    : never
+  : never
+
+export type ActionBaseType<
+  TChain extends Chain | undefined = Chain,
+  TChainOverride extends Chain | undefined = Chain | undefined,
+  _contractName extends string = string,
+  _resolvedChain extends Chain | undefined = ResolveChain<
+    TChain,
+    TChainOverride
+  >,
+  // TODO consider moving GetL2ChainId to make this even more generic
```

Comment:
i.e. we could move it out of this type and append it just in time in some places, or possibly we could simplify and do like `targetChain` instead of `L2ChainId`. But the name has to work for both reads and writes. I.e. `to` doesn't work cause in some cases we are specifying where we want to read data.

Reply:
Actually, in the case of L2 -> L1 actions we never pass the L1 chain ID. It should be known. So maybe this is fine. But we could update these types to have L1 and L2 base write actions